File: templates/_composition/CodeBehind/Project.SplitViewVB/Views/ShellNavigationItem.vb
Namespace Views
    Public NotInheritable Class ShellNavigationItem
        Implements INotifyPropertyChanged

        Public Property Label As String

        Public Property PageType As Type

        Public Property Symbol As Symbol

        Public ReadOnly Property SymbolAsChar As Char
            Get
                Return Convert.ToChar(Symbol)
            End Get
        End Property

        Private ReadOnly _iconElement As IconElement = Nothing

        Public ReadOnly Property Icon As IconElement
            Get
                Dim foregroundBinding = New Binding() With {
                    .Source = Me,
                    .Path = New PropertyPath("SelectedForeground"),
                    .Mode = BindingMode.OneWay,
                    .UpdateSourceTrigger = UpdateSourceTrigger.PropertyChanged
                }

                If _iconElement IsNot Nothing Then
                    BindingOperations.SetBinding(_iconElement, IconElement.ForegroundProperty, foregroundBinding)

                    Return _iconElement
                End If

                Dim fontIcon = New FontIcon() With {
                    .FontSize = 16,
                    .Glyph = SymbolAsChar.ToString()
                }

                BindingOperations.SetBinding(fontIcon, IconElement.ForegroundProperty, foregroundBinding)

                Return fontIcon
            End Get
        End Property

        Private _selectedVis As Visibility = Visibility.Collapsed

        Public Property SelectedVis As Visibility
            Get
                Return _selectedVis
            End Get
            Set
                [Set](_selectedVis, Value)
            End Set
        End Property

        Private _isSelected As Boolean

        Public Property IsSelected As Boolean
            Get
                Return _isSelected
            End Get
            Set
                [Set](_isSelected, Value)
                SelectedVis = If(Value, Visibility.Visible, Visibility.Collapsed)
                SelectedForeground = If(Value, TryCast(Application.Current.Resources("SystemControlForegroundAccentBrush"), SolidColorBrush), GetStandardTextColorBrush())
            End Set
        End Property

        Private Function GetStandardTextColorBrush() As SolidColorBrush
            Dim brush = TryCast(Application.Current.Resources("ThemeControlForegroundBaseHighBrush"), SolidColorBrush)

            Return brush
        End Function

        Private _selectedForeground As SolidColorBrush = Nothing

        Public Property SelectedForeground As SolidColorBrush
            Get
                If _selectedForeground Is Nothing Then
                    _selectedForeground = GetStandardTextColorBrush()
                End If

                Return _selectedForeground
            End Get
            Set
                [Set](_selectedForeground, Value)
            End Set
        End Property

        Private Sub New(label As String, symbol As Symbol, pageType As Type)
            Me.New(label, pageType)
            Me.Symbol = symbol
        End Sub

        Private Sub New(label As String, icon As IconElement, pageType As Type)
            Me.New(label, pageType)
            Me._iconElement = icon
        End Sub

        Private Sub New(label As String, pageType As Type)
            Me.Label = label
            Me.PageType = pageType
        End Sub

        Public Shared Function FromType(Of T As Page)(label As String, symbol As Symbol) As ShellNavigationItem
            Return New ShellNavigationItem(label, symbol, GetType(T))
        End Function

        Public Shared Function FromType(Of T As Page)(label As String, icon As IconElement) As ShellNavigationItem
            Return New ShellNavigationItem(label, icon, GetType(T))
        End Function

        Public Overrides Function ToString() As String
            Return Label
        End Function

        Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

        Private Sub [Set](Of T)(ByRef storage As T, value As T, <CallerMemberName> Optional propertyName As String = Nothing)
            If Equals(storage, value) Then
                Return
            End If

            storage = value
            OnPropertyChanged(propertyName)
        End Sub

        Private Sub OnPropertyChanged(propertyName As String)
            RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
        End Sub
    End Class
End Namespace

File: templates/_composition/CodeBehind/Project.SplitView/Views/ShellNavigationItem.cs
using System;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Data;
using Windows.UI.Xaml.Media;
using System.ComponentModel;
using System.Runtime.CompilerServices;

namespace wts.ItemName.Views
{
    public class ShellNavigationItem : INotifyPropertyChanged
    {
        public string Label { get; set; }
        public Symbol Symbol { get; set; }
        public Type PageType { get; set; }

        private Visibility _selectedVis = Visibility.Collapsed;
        public Visibility SelectedVis
        {
            get { return _selectedVis; }
            set { Set(ref _selectedVis, value); }
        }

        public char SymbolAsChar
        {
            get { return (char)Symbol; }
        }

        private IconElement _iconElement = null;
        public IconElement Icon
        {
            get
            {
                var foregroundBinding = new Binding
                {
                    Source = this,
                    Path = new PropertyPath("SelectedForeground"),
                    Mode = BindingMode.OneWay,
                    UpdateSourceTrigger = UpdateSourceTrigger.PropertyChanged
                };

                if (_iconElement != null)
                {
                    BindingOperations.SetBinding(_iconElement, IconElement.ForegroundProperty, foregroundBinding);

                    return _iconElement;
                }

                var fontIcon = new FontIcon { FontSize = 16, Glyph = SymbolAsChar.ToString() };

                BindingOperations.SetBinding(fontIcon, FontIcon.ForegroundProperty, foregroundBinding);

                return fontIcon;
            }
        }

        private bool _isSelected;
        public bool IsSelected
        {
            get
            {
                return _isSelected;
            }

            set
            {
                Set(ref _isSelected, value);
                
                SelectedVis = value ? Visibility.Visible : Visibility.Collapsed;

                SelectedForeground = IsSelected
                    ? Application.Current.Resources["SystemControlForegroundAccentBrush"] as SolidColorBrush
                    : GetStandardTextColorBrush();
            }
        }

        private SolidColorBrush _selectedForeground = null;
        public SolidColorBrush SelectedForeground
        {
            get { return _selectedForeground ?? (_selectedForeground = GetStandardTextColorBrush()); }
            set { Set(ref _selectedForeground, value); }
        }

        private ShellNavigationItem(string name, Symbol symbol, Type pageType) : this(name, pageType)
        {
            Symbol = symbol;
        }

        private ShellNavigationItem(string name, IconElement icon, Type pageType) : this(name, pageType)
        {
            _iconElement = icon;
        }

        private ShellNavigationItem(string name, Type pageType)
        {
            Label = name;
            PageType = pageType;
        }

        public static ShellNavigationItem FromType<T>(string name, Symbol symbol)
            where T : Page
        {
            return new ShellNavigationItem(name, symbol, typeof(T));
        }

        public static ShellNavigationItem FromType<T>(string name, IconElement icon)
            where T : Page
        {
            return new ShellNavigationItem(name, icon, typeof(T));
        }

        public override string ToString()
        {
            return Label;
        }

        private SolidColorBrush GetStandardTextColorBrush()
        {
            var brush = Application.Current.Resources["SystemControlForegroundBaseHighBrush"] as SolidColorBrush;

            return brush;
        }

        public event PropertyChangedEventHandler PropertyChanged;

        private void Set<T>(ref T storage, T value, [CallerMemberName]string propertyName = null)
        {
            if (Equals(storage, value))
            {
                return;
            }

            storage = value;
            OnPropertyChanged(propertyName);
        }

        private void OnPropertyChanged(string propertyName) => PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
    }
}
